py/client/pydeephaven/session.py

```diff
@@ -241,22 +272,29 @@ def exportable_objects(self) -> Dict[str, ticket_pb2.TypedTicket]:
 
     @property
     def grpc_metadata(self):
-        l = [(b'authorization', self._auth_token)]
+        header_value_snap = self._auth_header_value  # ensure it doesn't change while doing multiple reads
+        if header_value_snap is None or not isinstance(header_value_snap, bytes) or len(header_value_snap) == 0:
+            logger.warning(f'{self._logpfx} internal invariant violated, _auth_header_value={header_value_snap}')
+            l = []
+        else:
+            l = [(b'authorization', header_value_snap)]
         if self._extra_headers:
             l.extend(list(self._extra_headers.items()))
         return l
 
     @property
     def table_service(self) -> TableService:
-        if not self._table_service:
-            self._table_service = TableService(self)
-        return self._table_service
+        with self._r_lock:
+            if not self._table_service:
+                self._table_service = TableService(self)
+            return self._table_service
 
     @property
     def session_service(self) -> SessionService:
-        if not self._session_service:
-            self._session_service = SessionService(self)
-        return self._session_service
+        with self._r_lock:
+            if not self._session_service:
+                self._session_service = SessionService(self)
+            return self._session_service
```

Comment:
```suggestion
        if not self._session_service:
            with self._r_lock:
                if not self._session_service:
                    self._session_service = SessionService(self)
               return self._session_service
```
For better performance/concurrency and this is why I decided to just initialize them upfront in my PR.

Reply:
Done (in all of them, for consistency, and because if it is good/desired in one, should be good/desired for all)